grid_map_pcl/src/GridMapPclLoader.cpp

```diff
@@ -129,31 +129,50 @@ void GridMapPclLoader::processGridMapCell(const unsigned int linearGridMapIndex,
     return;
   }
   auto& clusterHeights = clusterHeightsWithingGridMapCell_[index(0)][index(1)];
-  calculateElevationFromPointsInsideGridMapCell(pointsInsideCellBorder, clusterHeights);
-  if (clusterHeights.empty()) {
-    (*gridMapData)(index(0), index(1)) = std::nan("1");
-  } else {
-    (*gridMapData)(index(0), index(1)) = params_.get().clusterExtraction_.useMaxHeightAsCellElevation_
-                                             ? *(std::max_element(clusterHeights.begin(), clusterHeights.end()))
-                                             : *(std::min_element(clusterHeights.begin(), clusterHeights.end()));
-  }
+  (*gridMapData)(index(0), index(1)) = calculateElevationFromPointsInsideGridMapCell(pointsInsideCellBorder, clusterHeights);
 }
 
-void GridMapPclLoader::calculateElevationFromPointsInsideGridMapCell(Pointcloud::ConstPtr cloud, std::vector<float>& heights) const {
+float GridMapPclLoader::calculateElevationFromPointsInsideGridMapCell(Pointcloud::ConstPtr cloud, std::vector<float>& heights) const {
   heights.clear();
   // Extract point cloud cluster from point cloud and return if none is found.
   std::vector<Pointcloud::Ptr> clusterClouds = pointcloudProcessor_.extractClusterCloudsFromPointcloud(cloud);
   const bool isNoClustersFound = clusterClouds.empty();
   if (isNoClustersFound) {
     ROS_WARN_STREAM_THROTTLE(10.0, "No clusters found in the grid map cell");
-    return;
+    return std::nan("1");
   }
 
   // Extract mean z value of cluster vector and return smallest height value
   heights.reserve(clusterClouds.size());
 
   std::transform(clusterClouds.begin(), clusterClouds.end(), std::back_inserter(heights),
                  [this](Pointcloud::ConstPtr cloud) -> double { return grid_map_pcl::calculateMeanOfPointPositions(cloud).z(); });
+
+  double height;
+  const int height_type = params_.get().gridMap_.height_type_;
+  // 0: smallest value among the mean values ​​of each cluster
+  if (height_type == 0) {
+    height = *(std::min_element(heights.begin(), heights.end()));
+  }
+  // 1: largest value among the mean values ​​of each cluster
+  else if (height_type == 1) {
+    height = *(std::max_element(heights.begin(), heights.end()));
+  }
+  // 2: Mean value of the cluster with the most points
+  else if (height_type == 2) {
+    const float min_height = *(std::min_element(heights.begin(), heights.end()));
+    std::vector<int> clusterSizes(clusterClouds.size());
+    for (size_t i = 0; i < clusterClouds.size(); i++) {
+      clusterSizes[i] = heights[i] - min_height < params_.get().gridMap_.height_thresh_ ?
+        clusterClouds[i]->size() :
+        -1;
+    }
+    const std::vector<int>::iterator maxIt =
+      std::max_element(clusterSizes.begin(), clusterSizes.end());
+    const size_t maxIndex = std::distance(clusterSizes.begin(), maxIt);
+    height = heights[maxIndex];
+  }
```

Comment:
Can you add an extra `else` for non-valid height options?

Reply:
Return std::nan and print error when `height_type` is invalid. https://github.com/ANYbotics/grid_map/pull/329/commits/0cc1b53ff7e544ed61a950106f514716fc2991ee

Reply:
Thank you.